### DISCUSSION TITLE
Allow File Accessibility Modifier On All Member Declarations

### DISCUSSION THREAD

**Discussion by TactiTac0z** [dotnet/csharplang#7196] [Language Ideas]
Right now the recently added 'file' accessibility modifier is valid for non-nested type declarations only.
In my opinion this seems like an unnecessary restriction.

Consider a simple example of a method that throws after a certain amount of times it has been called
```cs
partial class SomeBigClass
{
   int __callCount;
   public void TestMethod()
   {
      __callCount++;
      if (__callCount > 20)
         throw new Exception("Max call count hit");
   }
}
```
(Please don't get destracted by the exact example used, the example is not meant to be practile code in it self, but is merely meant to represent a simple scenario, I'm sure you can imagine other cases where you have a field (or any other member) you have no need nor intention of referencing anywhere else within the implementation of the type, like the backing store of a property for instance)

This works but is rarther ugly.
Don't forget our imaginay method is defined within a class that are potentually many files big.
We don't want to call the field something more natual like 'callCount' here, since that would pollute the scope of the rest of the class.
We use the "__" prefix to mean "uncommon" or "don't touch unless you know what you are doing" kind of nameing scheme as a substitute.

Lets look at an example using the 'file' modifier.
```cs
partial class SomeBigClass
{
   file int callCount;
   public void ThrowIfHighCallCount()
   {
     callCount++;
      if (callCount > 20)
         throw new Exception("Max call count hit");
   }
}
```

This looks a lot cleaner now.
We can use a much cleaner name for our field now AND not pollute the scope of the entire class with random fields.

Both consumers of this type (which already can't reference our field either way of course since it's private in the first example) 
AND other internal implementation files of `SomeBigClass` have no business in referencing `callCount`. So there is no need to have the field visible anywhere else then where we need it.

**Comment by RikkiGibson** [comment]
I want to do this. `file class` was really a tactical solution which allowed us to delay answering questions about how `file` interacts with existing accessibility on type members. For example:

```cs
class Class1
{
    file void M1() { }
}

class Class2
{
    void M2(Class1 c)
    {
        c.M1(); // is this allowed?
    }
}
```

**Reply by TactiTac0z** [reply]
I appreciate it.

The answer for that seems very simple.
M1 is allowed to be referenced there.
A member with the 'file' accessibility modifier should be declared as "internal" as far as the runtime (and other languages) is concerned.
But with an unspeakable name like `"<M1>_file_1"` or something like that, I believe the current implementation for types already does that.

like internal M1 is accessible to the hole assembly, but not more accessibly then it's parent member. 
BUT of course with the added syntax sugar that M1 is also only accessibly within the file it was declared.

So your example is allowed, but exactly like as if M1 was 'internal' or 'public'
```cs
class Class1
{
  class InnerClass
  {
    file static int x;
  }
}
class Class2
{
  void Test()
  {
    Class1.InnerClass.x++; // Accessibility violation. Cannot access private member 'InnerClass' here
  }
}
``` 
Is not allowed.
Because InnerClass is not accessibility as per usual nested accessibility rules.
But it would be allowed if InnerClass had an accessibility of file, internal, or public.

**Reply by FaustVX** [reply]
@RikkiGibson 
It depends if `file private/public/…` will be allowed in the future :
- If this will be allowed, then there is no accessibility modifiers, so it's `private` like other members, and your example is valid;
- Otherwise, `file` should be a new modifier, similar to `internal`, but limited to the whole file (like @TactiTac0z said), so your example is invalid.

**Reply by CyrusNajmabadi** [reply]
I would find it exceptionally odd for a child member to have internal accessibility by default.  I would expect these to have the same accessibility as if accessibility was not provided, but only available in the current file.

**Reply by FaustVX** [reply]
@CyrusNajmabadi 
If you won't be able to combine `file` with existing accessibility modifiers, then `file` should be a new one, accessible in the whole file.

(I have updated my previous comment)

**Reply by TactiTac0z** [reply]
Hmm I'm a bit confused @CyrusNajmabadi
I don't think anyone is talking about making anything internal by default.
unless you are talking about the "As far as the runtime is concerned bit" but that is more of a implementation detail.

I don't see 'file' as meaning something in addition to internal or private in a literal language design sense.
'file' as an accessibly level is a separate level. file is more accessibly then private and protected, but less accessibly then internal and public.
file void M();
has the accessibly of 'file', not file internal or file private or anything like that.

**Reply by CyrusNajmabadi** [reply]
I was responding to this:

> but limited to the whole file

I view it as a visibility modifier, not an accessibility one.  But this will certainly have to be hashed out.

**Reply by TactiTac0z** [reply]
Ohh ok that is fair.

If we do it where 'file' is a separate level entirely 
or an addition layer of restriction so my example would need 'x' to be 'file public' is ok with me :)

I would love it either way.

**Reply by FaustVX** [reply]
That's why I have split it in 2 points, the 1st is just like you said, as a visibility modifier, and the 2nd as a new accessibility modifier.
If it's a new one, it make sense to be accessible in the whole file, as it's name suggest.
Otherwise, we're ok, it's default accessibility is `private`

**Reply by TactiTac0z** [reply]
Yeah exactly like FaustVx says
